Wemos S2 Mini - I2S Iissue

### The problem

I am trying to use max98357a I2S Amplifier with Wemos S2 mini. According to the [Wemos S2 mini documentation](https://www.espressif.com/sites/default/files/documentation/esp32-s2_datasheet_en.pdf) I should be able to use any GPIO pins with I2S.
My ESPHome config:
```
media_player:
  - platform: i2s_audio
    name: ${friendly_name} Speaker
    dac_type: external
    i2s_lrclk_pin: 12
    i2s_dout_pin: 33
    i2s_bclk_pin: 11
    mode: mono
    id: media_out
```
but when I try to install it I get this:
```
INFO Reading configuration /config/esphome/szopa-multisensor.yaml...
WARNING The selected Arduino framework version is not the recommended one. If there are connectivity or build issues please remove the manual version.
WARNING The selected Arduino framework version is not the recommended one. If there are connectivity or build issues please remove the manual version.
INFO Generating C++ source...
INFO Compiling app...
Processing szopa-multisensor (board: lolin_s2_mini; framework: arduino; platform: platformio/espressif32 @ 4.4.0)
--------------------------------------------------------------------------------
HARDWARE: ESP32S2 240MHz, 320KB RAM, 4MB Flash
 - toolchain-riscv32-esp @ 8.4.0+2021r2-patch3 
 - toolchain-xtensa-esp32s2 @ 8.4.0+2021r2-patch3
LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
Dependency Graph
|-- WiFi @ 2.0.0
|-- ESPmDNS @ 2.0.0
|-- Update @ 2.0.0
|-- noise-c @ 0.1.4
|   |-- libsodium @ 1.10018.1
|-- WiFiClientSecure @ 2.0.0
|-- HTTPClient @ 2.0.0
|-- ESP32-audioI2S @ 2.1.0
Compiling /data/szopa-multisensor/.pioenvs/szopa-multisensor/src/esphome/components/i2s_audio/i2s_audio_media_player.cpp.o
Compiling /data/szopa-multisensor/.pioenvs/szopa-multisensor/src/esphome/core/component.cpp.o
Compiling /data/szopa-multisensor/.pioenvs/szopa-multisensor/src/esphome/core/component_iterator.cpp.o
Compiling /data/szopa-multisensor/.pioenvs/szopa-multisensor/src/esphome/core/controller.cpp.o
In file included from src/esphome/components/i2s_audio/i2s_audio_media_player.cpp:1:
src/esphome/components/i2s_audio/i2s_audio_media_player.h:28:30: error: 'i2s_dac_mode_t' has not been declared
   void set_internal_dac_mode(i2s_dac_mode_t mode) { this->internal_dac_mode_ = mode; }
                              ^~~~~~~~~~~~~~
src/esphome/components/i2s_audio/i2s_audio_media_player.h:54:3: error: 'i2s_dac_mode_t' does not name a type; did you mean 'i2s_mode_t'?
   i2s_dac_mode_t internal_dac_mode_{I2S_DAC_CHANNEL_DISABLE};
   ^~~~~~~~~~~~~~
   i2s_mode_t
src/esphome/components/i2s_audio/i2s_audio_media_player.h: In member function 'void esphome::i2s_audio::I2SAudioMediaPlayer::set_internal_dac_mode(int)':
src/esphome/components/i2s_audio/i2s_audio_media_player.h:28:59: error: 'class esphome::i2s_audio::I2SAudioMediaPlayer' has no member named 'internal_dac_mode_'; did you mean 'set_internal_dac_mode'?
   void set_internal_dac_mode(i2s_dac_mode_t mode) { this->internal_dac_mode_ = mode; }
                                                           ^~~~~~~~~~~~~~~~~~
                                                           set_internal_dac_mode
src/esphome/components/i2s_audio/i2s_audio_media_player.cpp: In member function 'virtual void esphome::i2s_audio::I2SAudioMediaPlayer::setup()':
src/esphome/components/i2s_audio/i2s_audio_media_player.cpp:106:13: error: 'class esphome::i2s_audio::I2SAudioMediaPlayer' has no member named 'internal_dac_mode_'; did you mean 'set_internal_dac_mode'?
   if (this->internal_dac_mode_ != I2S_DAC_CHANNEL_DISABLE) {
             ^~~~~~~~~~~~~~~~~~
             set_internal_dac_mode
src/esphome/components/i2s_audio/i2s_audio_media_player.cpp:106:35: error: 'I2S_DAC_CHANNEL_DISABLE' was not declared in this scope
   if (this->internal_dac_mode_ != I2S_DAC_CHANNEL_DISABLE) {
                                   ^~~~~~~~~~~~~~~~~~~~~~~
src/esphome/components/i2s_audio/i2s_audio_media_player.cpp:106:35: note: suggested alternative: 'PTHREAD_CANCEL_DISABLE'
   if (this->internal_dac_mode_ != I2S_DAC_CHANNEL_DISABLE) {
                                   ^~~~~~~~~~~~~~~~~~~~~~~
                                   PTHREAD_CANCEL_DISABLE
src/esphome/components/i2s_audio/i2s_audio_media_player.cpp:107:51: error: 'class esphome::i2s_audio::I2SAudioMediaPlayer' has no member named 'internal_dac_mode_'; did you mean 'set_internal_dac_mode'?
     this->audio_ = make_unique<Audio>(true, this->internal_dac_mode_);
                                                   ^~~~~~~~~~~~~~~~~~
                                                   set_internal_dac_mode
src/esphome/components/i2s_audio/i2s_audio_media_player.cpp: In member function 'virtual void esphome::i2s_audio::I2SAudioMediaPlayer::dump_config()':
src/esphome/components/i2s_audio/i2s_audio_media_player.cpp:140:13: error: 'class esphome::i2s_audio::I2SAudioMediaPlayer' has no member named 'internal_dac_mode_'; did you mean 'set_internal_dac_mode'?
   if (this->internal_dac_mode_ != I2S_DAC_CHANNEL_DISABLE) {
             ^~~~~~~~~~~~~~~~~~
             set_internal_dac_mode
src/esphome/components/i2s_audio/i2s_audio_media_player.cpp:140:35: error: 'I2S_DAC_CHANNEL_DISABLE' was not declared in this scope
   if (this->internal_dac_mode_ != I2S_DAC_CHANNEL_DISABLE) {
                                   ^~~~~~~~~~~~~~~~~~~~~~~
src/esphome/components/i2s_audio/i2s_audio_media_player.cpp:140:35: note: suggested alternative: 'PTHREAD_CANCEL_DISABLE'
   if (this->internal_dac_mode_ != I2S_DAC_CHANNEL_DISABLE) {
                                   ^~~~~~~~~~~~~~~~~~~~~~~
                                   PTHREAD_CANCEL_DISABLE
src/esphome/components/i2s_audio/i2s_audio_media_player.cpp:141:19: error: 'class esphome::i2s_audio::I2SAudioMediaPlayer' has no member named 'internal_dac_mode_'; did you mean 'set_internal_dac_mode'?
     switch (this->internal_dac_mode_) {
                   ^~~~~~~~~~~~~~~~~~
                   set_internal_dac_mode
src/esphome/components/i2s_audio/i2s_audio_media_player.cpp:142:12: error: 'I2S_DAC_CHANNEL_LEFT_EN' was not declared in this scope
       case I2S_DAC_CHANNEL_LEFT_EN:
            ^~~~~~~~~~~~~~~~~~~~~~~
src/esphome/components/i2s_audio/i2s_audio_media_player.cpp:142:12: note: suggested alternative: 'I2S_DATA_ENABLE_TEST_EN'
       case I2S_DAC_CHANNEL_LEFT_EN:
            ^~~~~~~~~~~~~~~~~~~~~~~
            I2S_DATA_ENABLE_TEST_EN
src/esphome/components/i2s_audio/i2s_audio_media_player.cpp:145:12: error: 'I2S_DAC_CHANNEL_RIGHT_EN' was not declared in this scope
       case I2S_DAC_CHANNEL_RIGHT_EN:
            ^~~~~~~~~~~~~~~~~~~~~~~~
src/esphome/components/i2s_audio/i2s_audio_media_player.cpp:145:12: note: suggested alternative: 'I2S_CHANNEL_FMT_RIGHT_LEFT'
       case I2S_DAC_CHANNEL_RIGHT_EN:
            ^~~~~~~~~~~~~~~~~~~~~~~~
            I2S_CHANNEL_FMT_RIGHT_LEFT
src/esphome/components/i2s_audio/i2s_audio_media_player.cpp:148:12: error: 'I2S_DAC_CHANNEL_BOTH_EN' was not declared in this scope
       case I2S_DAC_CHANNEL_BOTH_EN:
            ^~~~~~~~~~~~~~~~~~~~~~~
src/esphome/components/i2s_audio/i2s_audio_media_player.cpp:148:12: note: suggested alternative: 'I2S_CHANNEL_STEREO'
       case I2S_DAC_CHANNEL_BOTH_EN:
            ^~~~~~~~~~~~~~~~~~~~~~~
            I2S_CHANNEL_STEREO
Compiling /data/szopa-multisensor/.pioenvs/szopa-multisensor/src/esphome/core/entity_base.cpp.o
*** [/data/szopa-multisensor/.pioenvs/szopa-multisensor/src/esphome/components/i2s_audio/i2s_audio_media_player.cpp.o] Error 1
========================== [FAILED] Took 2.51 seconds ==========================
```

### Which version of ESPHome has the issue?

2022.10.2

### What type of installation are you using?

Home Assistant Add-on

### Which version of Home Assistant has the issue?

2022.10.2

### What platform are you using?

ESP32

### Board

Wemos S2 Mini

### Component causing the issue

I2S

### Example YAML snippet

```yaml
esphome:
  name: szopa-multisensor

esp32:
  board: lolin_s2_mini
  variant: esp32s2
  framework:
    type: arduino
    version: 2.0.3
    platform_version: 4.4.0

wifi:
<section removed for brevity>
logger:
api:
ota:
  password: <removed>

 
media_player:
  - platform: i2s_audio
    name: Speaker
    dac_type: external
    i2s_lrclk_pin: 12
    i2s_dout_pin: 33
    i2s_bclk_pin: 11
    mode: mono
    id: media_out
```


### Anything in the logs that might be useful for us?

_No response_

### Additional information

_No response_

Now that ESPHome supports Wemos S2 mini I updated the config with:

````
esp32:
  board: lolin_s2_mini
````
and I have re-tested it in esphome v2022.12.0 and HA v2022.12.5
No change still  failing to compile:
````
HARDWARE: ESP32S2 240MHz, 320KB RAM, 4MB Flash
 - toolchain-riscv32-esp @ 8.4.0+2021r2-patch3 
 - toolchain-xtensa-esp32s2 @ 8.4.0+2021r2-patch3
LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
Dependency Graph
|-- WiFi @ 2.0.0
|-- ESPmDNS @ 2.0.0
|-- Update @ 2.0.0
|-- noise-c @ 0.1.4
|   |-- libsodium @ 1.10018.1
|-- WiFiClientSecure @ 2.0.0
|-- HTTPClient @ 2.0.0
|-- ESP32-audioI2S @ 2.0.6
Compiling /data/szopa-multisensor/.pioenvs/szopa-multisensor/src/esphome/components/i2s_audio/i2s_audio_media_player.cpp.o
Compiling /data/szopa-multisensor/.pioenvs/szopa-multisensor/src/esphome/components/mdns/mdns_rp2040.cpp.o
Compiling /data/szopa-multisensor/.pioenvs/szopa-multisensor/src/esphome/components/media_player/media_player.cpp.o
Compiling /data/szopa-multisensor/.pioenvs/szopa-multisensor/src/esphome/components/network/util.cpp.o
Compiling /data/szopa-multisensor/.pioenvs/szopa-multisensor/src/esphome/components/ota/ota_backend_arduino_esp32.cpp.o
In file included from src/esphome/components/i2s_audio/i2s_audio_media_player.cpp:1:
src/esphome/components/i2s_audio/i2s_audio_media_player.h:28:30: error: 'i2s_dac_mode_t' has not been declared
   void set_internal_dac_mode(i2s_dac_mode_t mode) { this->internal_dac_mode_ = mode; }
                              ^~~~~~~~~~~~~~
src/esphome/components/i2s_audio/i2s_audio_media_player.h:54:3: error: 'i2s_dac_mode_t' does not name a type; did you mean 'i2s_mode_t'?
   i2s_dac_mode_t internal_dac_mode_{I2S_DAC_CHANNEL_DISABLE};
   ^~~~~~~~~~~~~~
   i2s_mode_t
src/esphome/components/i2s_audio/i2s_audio_media_player.h: In member function 'void esphome::i2s_audio::I2SAudioMediaPlayer::set_internal_dac_mode(int)':
src/esphome/components/i2s_audio/i2s_audio_media_player.h:28:59: error: 'class esphome::i2s_audio::I2SAudioMediaPlayer' has no member named 'internal_dac_mode_'; did you mean 'set_internal_dac_mode'?
   void set_internal_dac_mode(i2s_dac_mode_t mode) { this->internal_dac_mode_ = mode; }
                                                           ^~~~~~~~~~~~~~~~~~
                                                           set_internal_dac_mode
src/esphome/components/i2s_audio/i2s_audio_media_player.cpp: In member function 'virtual void esphome::i2s_audio::I2SAudioMediaPlayer::setup()':
src/esphome/components/i2s_audio/i2s_audio_media_player.cpp:106:13: error: 'class esphome::i2s_audio::I2SAudioMediaPlayer' has no member named 'internal_dac_mode_'; did you mean 'set_internal_dac_mode'?
   if (this->internal_dac_mode_ != I2S_DAC_CHANNEL_DISABLE) {
             ^~~~~~~~~~~~~~~~~~
             set_internal_dac_mode
src/esphome/components/i2s_audio/i2s_audio_media_player.cpp:106:35: error: 'I2S_DAC_CHANNEL_DISABLE' was not declared in this scope
   if (this->internal_dac_mode_ != I2S_DAC_CHANNEL_DISABLE) {
                                   ^~~~~~~~~~~~~~~~~~~~~~~
src/esphome/components/i2s_audio/i2s_audio_media_player.cpp:106:35: note: suggested alternative: 'PTHREAD_CANCEL_DISABLE'
   if (this->internal_dac_mode_ != I2S_DAC_CHANNEL_DISABLE) {
                                   ^~~~~~~~~~~~~~~~~~~~~~~
                                   PTHREAD_CANCEL_DISABLE
src/esphome/components/i2s_audio/i2s_audio_media_player.cpp:107:51: error: 'class esphome::i2s_audio::I2SAudioMediaPlayer' has no member named 'internal_dac_mode_'; did you mean 'set_internal_dac_mode'?
     this->audio_ = make_unique<Audio>(true, this->internal_dac_mode_);
                                                   ^~~~~~~~~~~~~~~~~~
                                                   set_internal_dac_mode
src/esphome/components/i2s_audio/i2s_audio_media_player.cpp: In member function 'virtual void esphome::i2s_audio::I2SAudioMediaPlayer::dump_config()':
src/esphome/components/i2s_audio/i2s_audio_media_player.cpp:140:13: error: 'class esphome::i2s_audio::I2SAudioMediaPlayer' has no member named 'internal_dac_mode_'; did you mean 'set_internal_dac_mode'?
   if (this->internal_dac_mode_ != I2S_DAC_CHANNEL_DISABLE) {
             ^~~~~~~~~~~~~~~~~~
             set_internal_dac_mode
src/esphome/components/i2s_audio/i2s_audio_media_player.cpp:140:35: error: 'I2S_DAC_CHANNEL_DISABLE' was not declared in this scope
   if (this->internal_dac_mode_ != I2S_DAC_CHANNEL_DISABLE) {
                                   ^~~~~~~~~~~~~~~~~~~~~~~
Compiling /data/szopa-multisensor/.pioenvs/szopa-multisensor/src/esphome/components/ota/ota_backend_arduino_esp8266.cpp.o
src/esphome/components/i2s_audio/i2s_audio_media_player.cpp:140:35: note: suggested alternative: 'PTHREAD_CANCEL_DISABLE'
   if (this->internal_dac_mode_ != I2S_DAC_CHANNEL_DISABLE) {
                                   ^~~~~~~~~~~~~~~~~~~~~~~
                                   PTHREAD_CANCEL_DISABLE
src/esphome/components/i2s_audio/i2s_audio_media_player.cpp:141:19: error: 'class esphome::i2s_audio::I2SAudioMediaPlayer' has no member named 'internal_dac_mode_'; did you mean 'set_internal_dac_mode'?
     switch (this->internal_dac_mode_) {
                   ^~~~~~~~~~~~~~~~~~
                   set_internal_dac_mode
src/esphome/components/i2s_audio/i2s_audio_media_player.cpp:142:12: error: 'I2S_DAC_CHANNEL_LEFT_EN' was not declared in this scope
       case I2S_DAC_CHANNEL_LEFT_EN:
            ^~~~~~~~~~~~~~~~~~~~~~~
Compiling /data/szopa-multisensor/.pioenvs/szopa-multisensor/src/esphome/components/ota/ota_backend_arduino_rp2040.cpp.o
src/esphome/components/i2s_audio/i2s_audio_media_player.cpp:142:12: note: suggested alternative: 'I2S_DATA_ENABLE_TEST_EN'
       case I2S_DAC_CHANNEL_LEFT_EN:
            ^~~~~~~~~~~~~~~~~~~~~~~
            I2S_DATA_ENABLE_TEST_EN
src/esphome/components/i2s_audio/i2s_audio_media_player.cpp:145:12: error: 'I2S_DAC_CHANNEL_RIGHT_EN' was not declared in this scope
       case I2S_DAC_CHANNEL_RIGHT_EN:
            ^~~~~~~~~~~~~~~~~~~~~~~~
Compiling /data/szopa-multisensor/.pioenvs/szopa-multisensor/src/esphome/components/ota/ota_backend_esp_idf.cpp.o
src/esphome/components/i2s_audio/i2s_audio_media_player.cpp:145:12: note: suggested alternative: 'I2S_CHANNEL_FMT_RIGHT_LEFT'
       case I2S_DAC_CHANNEL_RIGHT_EN:
            ^~~~~~~~~~~~~~~~~~~~~~~~
            I2S_CHANNEL_FMT_RIGHT_LEFT
src/esphome/components/i2s_audio/i2s_audio_media_player.cpp:148:12: error: 'I2S_DAC_CHANNEL_BOTH_EN' was not declared in this scope
       case I2S_DAC_CHANNEL_BOTH_EN:
            ^~~~~~~~~~~~~~~~~~~~~~~
Compiling /data/szopa-multisensor/.pioenvs/szopa-multisensor/src/esphome/components/ota/ota_component.cpp.o
Compiling /data/szopa-multisensor/.pioenvs/szopa-multisensor/src/esphome/components/sensor/automation.cpp.o
src/esphome/components/i2s_audio/i2s_audio_media_player.cpp:148:12: note: suggested alternative: 'I2S_CHANNEL_STEREO'
       case I2S_DAC_CHANNEL_BOTH_EN:
            ^~~~~~~~~~~~~~~~~~~~~~~
            I2S_CHANNEL_STEREO
Compiling /data/szopa-multisensor/.pioenvs/szopa-multisensor/src/esphome/components/sensor/filter.cpp.o
*** [/data/szopa-multisensor/.pioenvs/szopa-multisensor/src/esphome/components/i2s_audio/i2s_audio_media_player.cpp.o] Error 1
========================== [FAILED] Took 2.70 seconds ==========================
````

--- a/esphome/components/i2s_audio/i2s_audio_media_player.cpp
+++ b/esphome/components/i2s_audio/i2s_audio_media_player.cpp
@@ -103,9 +103,11 @@ void I2SAudioMediaPlayer::stop_() {
 
 void I2SAudioMediaPlayer::setup() {
   ESP_LOGCONFIG(TAG, "Setting up Audio...");
+#if SOC_I2S_SUPPORTS_DAC
   if (this->internal_dac_mode_ != I2S_DAC_CHANNEL_DISABLE) {
     this->audio_ = make_unique<Audio>(true, this->internal_dac_mode_);
   } else {
+#endif
     this->audio_ = make_unique<Audio>(false);
     this->audio_->setPinout(this->bclk_pin_, this->lrclk_pin_, this->dout_pin_);
     this->audio_->forceMono(this->external_dac_channels_ == 1);
@@ -113,7 +115,9 @@ void I2SAudioMediaPlayer::setup() {
       this->mute_pin_->setup();
       this->mute_pin_->digital_write(false);
     }
+#if SOC_I2S_SUPPORTS_DAC
   }
+#endif
   this->state = media_player::MEDIA_PLAYER_STATE_IDLE;
 }
 
@@ -137,6 +141,7 @@ void I2SAudioMediaPlayer::dump_config() {
     ESP_LOGCONFIG(TAG, "Audio failed to initialize!");
     return;
   }
+#if SOC_I2S_SUPPORTS_DAC
   if (this->internal_dac_mode_ != I2S_DAC_CHANNEL_DISABLE) {
     switch (this->internal_dac_mode_) {
       case I2S_DAC_CHANNEL_LEFT_EN:
@@ -152,6 +157,7 @@ void I2SAudioMediaPlayer::dump_config() {
         break;
     }
   }
+#endif
 }
 
 }  // namespace i2s_audio

--- a/esphome/components/i2s_audio/i2s_audio_media_player.h
+++ b/esphome/components/i2s_audio/i2s_audio_media_player.h
@@ -25,7 +25,9 @@ class I2SAudioMediaPlayer : public Component, public media_player::MediaPlayer {
   void set_bclk_pin(uint8_t pin) { this->bclk_pin_ = pin; }
   void set_lrclk_pin(uint8_t pin) { this->lrclk_pin_ = pin; }
   void set_mute_pin(GPIOPin *mute_pin) { this->mute_pin_ = mute_pin; }
+#if SOC_I2S_SUPPORTS_DAC
   void set_internal_dac_mode(i2s_dac_mode_t mode) { this->internal_dac_mode_ = mode; }
+#endif
   void set_external_dac_channels(uint8_t channels) { this->external_dac_channels_ = channels; }
 
   media_player::MediaPlayerTraits get_traits() override;
@@ -51,7 +53,9 @@ class I2SAudioMediaPlayer : public Component, public media_player::MediaPlayer {
   bool muted_{false};
   float unmuted_volume_{0};
 
+#if SOC_I2S_SUPPORTS_DAC
   i2s_dac_mode_t internal_dac_mode_{I2S_DAC_CHANNEL_DISABLE};
+#endif
   uint8_t external_dac_channels_;
 
   HighFrequencyLoopRequester high_freq_;

--- a/esphome/components/i2s_audio/media_player.py
+++ b/esphome/components/i2s_audio/media_player.py
@@ -1,5 +1,5 @@
 import esphome.codegen as cg
-from esphome.components import media_player
+from esphome.components import media_player, esp32
 import esphome.config_validation as cv
 
 from esphome import pins
@@ -32,6 +32,18 @@ INTERNAL_DAC_OPTIONS = {
 }
 
 EXTERNAL_DAC_OPTIONS = ["mono", "stereo"]
+
+NO_INTERNAL_DAC_VARIANTS = [esp32.const.VARIANT_ESP32S2]
+
+
+def validate_esp32_variant(config):
+    if config[CONF_DAC_TYPE] != "internal":
+        return config
+    variant = esp32.get_esp32_variant()
+    if variant in NO_INTERNAL_DAC_VARIANTS:
+        raise cv.Invalid(f"{variant} does not have an internal DAC")
+    return config
+
 
 CONFIG_SCHEMA = cv.All(
     cv.typed_schema(
@@ -68,6 +80,7 @@ CONFIG_SCHEMA = cv.All(
         key=CONF_DAC_TYPE,
     ),
     cv.only_with_arduino,
+    validate_esp32_variant,
 )
 
 

